# HG changeset patch
# User Pantxo Diribarne <pantxo.diribarne@gmail.com>
# Date 1396637392 -7200
#      Fri Apr 04 20:49:52 2014 +0200
# Node ID ee125c82e01e9c78187b20053b7cacfc102f69d6
# Parent  37d5a2bb4160304e20446d762adc74f3caf8556b
struct2hdl.m: handle properly objects with non-empy "tag" (bug #42047).

* struct2hdl.m: don't exclude object properties with non-empty "tag".

* copyobj.m: add a test for bug #42047.

diff --git a/scripts/plot/util/copyobj.m b/scripts/plot/util/copyobj.m
--- a/scripts/plot/util/copyobj.m
+++ b/scripts/plot/util/copyobj.m
@@ -163,8 +163,22 @@ endfunction
 %!   assert (map1, map2);
 %!   assert (alpha1, alpha2);
 %! unwind_protect_cleanup
 %!   close (h1);
 %!   close (h2);
 %!   graphics_toolkit (toolkit);
 %! end_unwind_protect
 
+%!test 
+%! unwind_protect
+%!   tag = "foo";
+%!   hf = figure ("visible", "off");
+%!   hax = axes ("tag", tag);
+%!   hpa = patch ();
+%!   set (hpa, "facecolor", [.5 .5 .5], "tag", tag)
+%!   hax2 = copyobj (hax, hf);
+%!   assert (get (hax2, "tag"), tag)
+%!   hpa2 = get (hax2, "children");
+%!   assert (get (hpa2, "facecolor"), [.5 .5 .5])
+%! unwind_protect_cleanup
+%!   close (hf)
+%! end_unwind_protect
diff --git a/scripts/plot/util/struct2hdl.m b/scripts/plot/util/struct2hdl.m
--- a/scripts/plot/util/struct2hdl.m
+++ b/scripts/plot/util/struct2hdl.m
@@ -172,17 +172,17 @@ function [h, pout] = struct2hdl (s, p=[]
   setprops (s, h, p, hilev);
 
   pout = p;
 
 endfunction
 
 function [h, sout] = createaxes (s, p, par)
   ## regular axes
-  if (strcmp (s.properties.tag, ""))
+  if (! any (strcmpi (s.properties.tag, {"colorbar", "legend"})))
     propval = {"position", s.properties.position};
     hid = {"autopos_tag", "looseinset"};
     for ii = 1:numel (hid)
       prop = hid{ii};
       if (isfield (s.properties, prop))
         val = s.properties.(prop);
         propval = [propval, prop, val];
       endif
@@ -546,18 +546,17 @@ function [h, sout, pout] = createhg_hile
     addmissingprops (h, s.properties);
     s.special = [];           # children will be treated as normal children
   endif
   sout = s;
   pout = p;
 endfunction
 
 function setprops (s, h, p, hilev)
-  more off;
-  if (strcmpi (s.properties.tag, ""))
+  if (! any (strcmpi (s.properties.tag, {"colorbar", "legend"})))
     specs = s.children(s.special);
     if (isempty (specs))
       hdls = [];
     else
       hdls = [specs.handle];
     endif
     nh = length (hdls);
     msg = "";
@@ -612,18 +611,17 @@ function setprops (s, h, p, hilev)
 
           else
             error ("struct2hdl: non-conformant number of children in hgggroup");
           endif
         endif
       endif
     endif
 
-  elseif (strcmpi (s.properties.tag, "legend")
-          || strcmpi (s.properties.tag, "colorbar"))
+  else
     set (h, s.properties);
   endif
 
 endfunction
 
 function out = valcomp (x, hdls)
   if (isfloat (x) && isscalar (x))
     out = any (x == hdls);
